### PR TITLE
Add gradient computations in reverse mode

### PR DESCRIPTION
**Pull request by efremale**
For a function ```f```, its gradient can be computed by calling ```clad::gradient(f)```.

For example, calling ```clad::gradient``` on
```
double rosenbrock_func(double x, double y) {
  return (x - 1) * (x - 1) + 100 * (y - x * x) * (y - x * x);
}
```
should generate
```
void rosenbrock_func_grad(double x, double y, double *_result) {
    _result[0UL] += 1. * (x - 1);
    _result[0UL] += 1. * (x - 1);
    _result[1UL] += 1. * (y - x * x) * 100;
    _result[0UL] += -1. * (y - x * x) * 100 * x;
    _result[0UL] += -1. * (y - x * x) * 100 * x;
    _result[1UL] += 1. * 100 * (y - x * x);
    _result[0UL] += -1. * 100 * (y - x * x) * x;
    _result[0UL] += -1. * 100 * (y - x * x) * x;
}
```

The caller of the ```*_grad``` functions is responsible for allocating enough space for the ```_result``` array.

### CODE REVIEW

File: lib/Differentiator/GradientBuilder.cpp [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+      VK_LValue,
+      noLoc()).get();
+    //initially df/df = 1
+    auto dfdf = ConstantFolder::synthesizeLiteral(FD->getReturnType(), m_Context, 1.0);
+    //this starts the visitation process and accumulates the statements in the m_BodyStmts
+    visitWith(FD->getMostRecentDecl()->getBody(), dfdf);
+
+    llvm::ArrayRef<Stmt*> stmtsRef =
+      llvm::makeArrayRef(m_BodyStmts.data(), m_BodyStmts.size());
+    //create the body of the function
```
Comment (vgvassilev) [inline lib/Differentiator/GradientBuilder.cpp]: Likewise.

File: include/clad/Differentiator/GradientBuilder.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    std::unique_ptr<utils::StmtClone> m_NodeCloner;
+    clang::NamespaceDecl* m_BuiltinDerivativesNSD;
+    bool m_GradientInFlight;
+
+    //stack used to pass the arguments (dfdx) to further nodes in Visit
+    std::stack<clang::Expr*> m_Stack;
+    clang::Expr* dfdx () {
+        return m_Stack.top ();
+    }
+    void visitWith(const clang::Stmt* stmt, clang::Expr* expr) {
```
Comment (vgvassilev) [inline include/clad/Differentiator/GradientBuilder.h]: Can we not override just ::Visit?

File: lib/Differentiator/GradientBuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+#include "clang/Sema/Overload.h"
+#include "clang/Sema/SemaInternal.h"
+
+#include <algorithm>
+
+using namespace clang;
+
+
+namespace clad {
+  //same as in the Derivative Builder
```
Comment (vgvassilev) [inline lib/Differentiator/GradientBuilder.cpp]: Could you please merge them and parameterize the different parts?

File: lib/Differentiator/GradientBuilder.cpp [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    Sema::SynthesizedFunctionScope Scope(m_Sema, gradientFD);
+    //ref to the last parameter
+    m_Result = m_Sema.BuildDeclRefExpr(
+      params.back(),
+      paramTypes.back(),
+      VK_LValue,
+      noLoc()).get();
+    //initially df/df = 1
+    auto dfdf = ConstantFolder::synthesizeLiteral(FD->getReturnType(), m_Context, 1.0);
+    //this starts the visitation process and accumulates the statements in the m_BodyStmts
```
[the remark this answers is not in the document]
Comment (efremale) [reply lib/Differentiator/GradientBuilder.cpp]: Fixed

File: lib/Differentiator/GradientBuilder.cpp (hunk 1 above)
Comment (efremale) [reply lib/Differentiator/GradientBuilder.cpp]: Fixed.

File: include/clad/Differentiator/GradientBuilder.h (hunk 2 above)
Comment (efremale) [reply include/clad/Differentiator/GradientBuilder.h]: Yes, changed visitWith to Visit.

File: include/Math/CladDifferentiator.h [hunk 5]
```diff
@@ -0,0 +1,200 @@
+//--------------------------------------------------------------------*- C++ -*-
```
[the remark this answers is not in the document]
Comment (efremale) [reply include/Math/CladDifferentiator.h]: I decided to revert the changes and remove Math/CladDifferentiator.h and keep Differentiator.h and BuiltinDerivatives.h.

File: lib/Differentiator/GradientBuilder.cpp [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+      noLoc(),
+      name,
+      gradientFunctionType,
+      FD->getTypeSourceInfo(), //what is TypeSourceInfo for?
+      FD->getStorageClass(),
+      FD->isInlineSpecified(),
+      FD->hasWrittenPrototype(),
+      FD->isConstexpr());
+
+    m_CurScope.reset(new Scope(m_Sema.TUScope, Scope::FnScope,
```
[the remark this answers is not in the document]
Comment (efremale) [reply lib/Differentiator/GradientBuilder.cpp]: I am not sure about it for now, but it was used in DerivativeBuilder and it seems to be relevant.

File: include/clad/Differentiator/BuiltinDerivatives.h [hunk 7]
```diff
@@ -1,9 +1,3 @@
-//--------------------------------------------------------------------*- C++ -*-
```
Comment (vgvassilev) [inline include/clad/Differentiator/BuiltinDerivatives.h]: Could you bring this back?

File: include/clad/Differentiator/Differentiator.h [hunk 8]
```diff
@@ -1,9 +1,3 @@
-//--------------------------------------------------------------------*- C++ -*-
```
Comment (vgvassilev) [inline include/clad/Differentiator/Differentiator.h]: Likewise.

File: test/FirstDerivative/DiffInterface.C [hunk 9]
```diff
@@ -51,8 +51,8 @@ int main () {
   clad::differentiate(f_2, 2);
 
   clad::differentiate(f_2, -1); // expected-error {{Invalid argument index -1 among 3 argument(s)}}
-  // expected-note@clad/Differentiator/Differentiator.h:97 {{candidate function not viable: no known conversion from 'int (int, float, int)' to 'unsigned int' for 2nd argument}}
-  // expected-note@clad/Differentiator/Differentiator.h:104 {{candidate template ignored: could not match 'R (C::*)(Args...)' against 'int (*)(int, float, int)'}}
+  // expected-note@clad/Differentiator/Differentiator.h:91 {{candidate function not viable: no known conversion from 'int (int, float, int)' to 'unsigned int' for 2nd argument}}
```
Comment (vgvassilev) [inline test/FirstDerivative/DiffInterface.C]: If you fix the header file comments you should not need to change this one.

File: test/FirstDerivative/DiffInterface.C [hunk 10]
```diff
@@ -51,8 +51,8 @@ int main () {
...
+  // expected-note@clad/Differentiator/Differentiator.h:98 {{candidate template ignored: could not match 'R (C::*)(Args...)' against 'int (*)(int, float, int)'}}
```
Comment (vgvassilev) [inline test/FirstDerivative/DiffInterface.C]: Likewise.

File: tools/ClangPlugin.cpp [hunk 11]
```diff
@@ -148,13 +159,76 @@ namespace clad {
                Derivative->print(f, Policy);
                f.flush();
             }
+
```
Comment (vgvassilev) [inline tools/ClangPlugin.cpp]: Please remove this whitespace.

File: include/clad/Differentiator/Differentiator.h [hunk 12]
```diff
@@ -105,5 +99,26 @@ namespace clad {
...
+  using grad_ptr_t = void (*)(Args..., R*);
+
+  template <typename Ptr, typename R, typename ... Args>
+  CladFunction<false, void, Args..., R*> __attribute__((annotate("GR")))
+  make_gradient(Ptr f, const char * code = "") {
+    return CladFunction<false, void, Args..., R*>(f, code);
+  }
+
+  template <typename ... Args>
+  void dummy(Args ...) {}
```
Comment (vgvassilev) [inline include/clad/Differentiator/Differentiator.h]: I'd recommend to call it `result_placeholder`.

File: include/clad/Differentiator/GradientBuilder.h [hunk 13]
```diff
@@ -0,0 +1,101 @@
+//--------------------------------------------------------------------*- C++ -*-
+// clad - the C++ Clang-based Automatic Differentiator
+// version: $Id: ClangPlugin.cpp 7 2013-06-01 22:48:03Z v.g.vassilev@gmail.com $
+// author:  Vassil Vassilev <vvasilev-at-cern.ch>
```
Comment (vgvassilev) [inline include/clad/Differentiator/GradientBuilder.h]: You can put your name if you want.

File: include/clad/Differentiator/GradientBuilder.h [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  private:
+    clang::Sema& m_Sema;
+    clang::ASTContext& m_Context;
+    clang::ValueDecl* m_IndependentVar;
+    std::unique_ptr<clang::Scope> m_CurScope;
+    std::unique_ptr<utils::StmtClone> m_NodeCloner;
+    clang::NamespaceDecl* m_BuiltinDerivativesNSD;
+    bool m_GradientInFlight;
+
+    // Stack is used to pass the arguments (dfdx) to further nodes in the Visit method.
```
Comment (vgvassilev) [inline include/clad/Differentiator/GradientBuilder.h]: Please use `///` for doxygen to kick in.

File: include/clad/Differentiator/GradientBuilder.h [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    clang::Expr* m_Result;
+
+    void updateReferencesOf(clang::Stmt* InSubtree);
+
+    clang::SourceLocation noLoc() const;
+
+    clang::Stmt* Clone(const clang::Stmt* S);
+    clang::Expr* Clone(const clang::Expr* E);
+
+    clang::Expr* op(clang::UnaryOperatorKind opcode, clang::Expr* e);
```
Comment (vgvassilev) [inline include/clad/Differentiator/GradientBuilder.h]: Can you explain a little more the purpose of this and maybe change the name into something more meaningful?

File: include/clad/Differentiator/GradientBuilder.h [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    void VisitDeclStmt(const clang::DeclStmt* DS);
+    void VisitStmt(const clang::Stmt* S);
+    void VisitMemberExpr(const clang::MemberExpr* ME);
+    void VisitDeclRefExpr(const clang::DeclRefExpr* DRE);
+    void VisitParenExpr(const clang::ParenExpr* PE);
+    void VisitIntegerLiteral(const clang::IntegerLiteral* IL);
+    void VisitFloatingLiteral(const clang::FloatingLiteral* FL);
+    void VisitCallExpr(const clang::CallExpr* CE);
+    void VisitImplicitCastExpr(const clang::ImplicitCastExpr* ICE);
+
```
Comment (vgvassilev) [inline include/clad/Differentiator/GradientBuilder.h]: Extra blanks.

File: lib/Differentiator/DiffPlanner.cpp [hunk 17]
```diff
@@ -6,6 +6,9 @@
 using namespace clang;
 
 namespace clad {
+  //ugly workaround
```
Comment (vgvassilev) [inline lib/Differentiator/DiffPlanner.cpp]: A `FIXME: ` plus explanation why this is an ugly workaround and how we should make it less ugly ;)

File: lib/Differentiator/GradientBuilder.cpp (hunk 3 above)
Comment (efremale) [reply lib/Differentiator/GradientBuilder.cpp]: Done (at least partially, now we have a single DerivativeBuilder that calls visitors ForwardModeVisitor or ReverseModeVisitor).

File: include/clad/Differentiator/BuiltinDerivatives.h (hunk 7 above)
Comment (efremale) [reply include/clad/Differentiator/BuiltinDerivatives.h]: Done.

File: test/FirstDerivative/DiffInterface.C (hunk 9 above)
Comment (efremale) [reply test/FirstDerivative/DiffInterface.C]: Fixed.

File: test/FirstDerivative/DiffInterface.C (hunk 10 above)
Comment (efremale) [reply test/FirstDerivative/DiffInterface.C]: Fixed.

File: tools/ClangPlugin.cpp (hunk 11 above)
Comment (efremale) [reply tools/ClangPlugin.cpp]: Fixed.

File: include/clad/Differentiator/Differentiator.h (hunk 12 above)
Comment (efremale) [reply include/clad/Differentiator/Differentiator.h]: Renamed.

File: include/clad/Differentiator/GradientBuilder.h (hunk 14 above)
Comment (efremale) [reply include/clad/Differentiator/GradientBuilder.h]: Fixed.

File: include/clad/Differentiator/GradientBuilder.h (hunk 15 above)
Comment (efremale) [reply include/clad/Differentiator/GradientBuilder.h]: Added a comment, now these functions are just for better code readability.

File: lib/Differentiator/DiffPlanner.cpp (hunk 17 above)
Comment (efremale) [reply lib/Differentiator/DiffPlanner.cpp]: Fixed.

File: include/clad/Differentiator/BuiltinDerivatives.h [hunk 18]
```diff
@@ -78,5 +78,4 @@ namespace custom_derivatives {
   }
 
 } // end namespace builtin_derivatives
-
-#endif //CLAD_BUILTIN_DERIVATIVES
+#endif // CLAD_BUILTIN_DERIVATIVES
```
Comment (vgvassilev) [inline include/clad/Differentiator/BuiltinDerivatives.h]: Please readd the blankspace. This should not appear in the diff.

File: include/clad/Differentiator/DerivativeBuilder.h [hunk 19]
```diff
@@ -79,20 +82,32 @@ namespace clad {
   };
 
   class DiffPlan;
-  class DerivativeBuilder
-    : public clang::ConstStmtVisitor<DerivativeBuilder, NodeContext> {
+  /// The main builder class which then uses either
```
Comment (vgvassilev) [inline include/clad/Differentiator/DerivativeBuilder.h]: All code and comments should be aligned to 80 cols.

File: include/clad/Differentiator/DerivativeBuilder.h [hunk 20]
```diff
@@ -79,20 +82,32 @@ namespace clad {
...
 
+    /// Updates references in newly cloned statements.
     void updateReferencesOf(clang::Stmt* InSubtree);
+    /// Clones a statement
+    clang::Stmt* Clone(const clang::Stmt* S);
+    /// A shorthand to simplify cloning of expressions.
+    clang::Expr* Clone(const clang::Expr* E);
+    /// A shorthand to simplify syntax for creation of new expressions.
+    /// Uses m_Sema.BuildUnOp internally.
+    clang::Expr* Op(clang::UnaryOperatorKind opcode, clang::Expr* e);
```
Comment (vgvassilev) [inline include/clad/Differentiator/DerivativeBuilder.h]: Could you capitalize the `opcode`->`OpCode` and `e`->`E`? What about `Op`->`BuildOp`?

File: include/clad/Differentiator/DerivativeBuilder.h [hunk 21]
```diff
@@ -79,20 +82,32 @@ namespace clad {
...
+    /// Uses m_Sema.BuildBin internally.
+    clang::Expr* Op(clang::BinaryOperatorKind opcode, clang::Expr* l, clang::Expr* r);
```
Comment (vgvassilev) [inline include/clad/Differentiator/DerivativeBuilder.h]: Likewise.

File: include/clad/Differentiator/StmtClone.h [hunk 22]
```diff
@@ -131,7 +135,16 @@ namespace utils {
     return static_cast<StmtTy*>(clonedStmt);
   }
 
-} // namespace ASTProcessing
-} // namespace clang
+  class ReferencesUpdater : public clang::RecursiveASTVisitor<ReferencesUpdater> {
+  private:
+    clang::Sema& m_Sema; // We don't own.
```
Comment (vgvassilev) [inline include/clad/Differentiator/StmtClone.h]: This is passed by ref, the comment is redundant.

File: tools/ClangPlugin.cpp [hunk 23]
```diff
@@ -125,34 +124,34 @@ namespace clad {
               SimpleTimer Timer(WantTiming);
               Timer.setOutput("Generation time for "
                               + plan->begin()->getFD()->getNameAsString());
-
-              Derivative = m_DerivativeBuilder->Derive(*I, plan);
+
```
Comment (vgvassilev) [inline tools/ClangPlugin.cpp]: Blanks.

File: tools/ClangPlugin.cpp [hunk 24]
```diff
@@ -125,34 +124,34 @@ namespace clad {
...
             if (m_DO.GenerateSourceFile) {
                std::error_code err;
                llvm::raw_fd_ostream f("Derivatives.cpp", err,
                                       llvm::sys::fs::F_Append);
-               Derivative->print(f, Policy);
+               Result->print(f, Policy);
                f.flush();
-            }
-            if (Derivative) {
+            }
```
Comment (vgvassilev) [inline tools/ClangPlugin.cpp]: More blanks.

File: include/clad/Differentiator/BuiltinDerivatives.h (hunk 18 above)
Comment (efremale) [reply include/clad/Differentiator/BuiltinDerivatives.h]: Fixed.

File: include/clad/Differentiator/DerivativeBuilder.h (hunk 19 above)
Comment (efremale) [reply include/clad/Differentiator/DerivativeBuilder.h]: Done, though some old code is still a bit longer that 80 cols.

File: include/clad/Differentiator/DerivativeBuilder.h (hunk 20 above)
Comment (efremale) [reply include/clad/Differentiator/DerivativeBuilder.h]: Done.

File: include/clad/Differentiator/DerivativeBuilder.h (hunk 21 above)
Comment (efremale) [reply include/clad/Differentiator/DerivativeBuilder.h]: Changed.

File: include/clad/Differentiator/StmtClone.h (hunk 22 above)
Comment (efremale) [reply include/clad/Differentiator/StmtClone.h]: I did not add it, it was in Updater before.

File: include/clad/Differentiator/DiffPlanner.h [hunk 25]
```diff
@@ -43,12 +59,18 @@ namespace clad {
                  m_CurrentDerivativeOrder(1), m_ArgIndex(0) { }
     typedef Functions::iterator iterator;
     typedef Functions::const_iterator const_iterator;
-    unsigned getRequestedDerivativeOrder() { return m_RequestedDerivativeOrder;}
+
+    DiffMode mode;
```
[the remark this answers is not in the document]
Comment (efremale) [reply include/clad/Differentiator/DiffPlanner.h]: Yes, though I do not like the idea of getters and setters for everything.

**Pull request merged**